pkg/controller/datavolume-controller.go

```diff
@@ -2565,3 +2578,81 @@ func GetRequiredSpace(filesystemOverhead float64, requestedSpace int64) int64 {
 func newLongTermCloneTokenGenerator(key *rsa.PrivateKey) token.Generator {
 	return token.NewGenerator(common.ExtendedCloneTokenIssuer, key, 10*365*24*time.Hour)
 }
+
+func (r *DatavolumeReconciler) garbageCollect(dataVolume *cdiv1.DataVolume, pvc *corev1.PersistentVolumeClaim) (res *reconcile.Result, err error) {
+	if dataVolume.Status.Phase != cdiv1.Succeeded {
+		return
+	}
+	cdiConfig := &cdiv1.CDIConfig{}
+	if err = r.client.Get(context.TODO(), types.NamespacedName{Name: common.ConfigName}, cdiConfig); err != nil {
+		return
+	}
+	dvTTL := cdiConfig.Spec.DataVolumeTTLSeconds
+	if dvTTL == nil || *dvTTL < 0 {
+		return
+	}
+
+	// Current DV still has TTL, so reconcile will return with the needed RequeueAfter
+	if delta := getDeltaTTL(dataVolume, *dvTTL); delta > 0 {
+		return &reconcile.Result{RequeueAfter: delta}, nil
+	}
+
+	if err = r.detachPvcDeleteDv(pvc, dataVolume); err != nil {
+		return
+	}
+
+	// FIXME: not sure this part is necessary
+	// Lookup for other DVs to garbage collect
+	selector := fields.OneTermEqualSelector(dvPhaseField, string(cdiv1.Succeeded))
+	dvList := &cdiv1.DataVolumeList{}
+	if err = r.client.List(context.TODO(), dvList, &client.ListOptions{FieldSelector: selector}); err != nil {
+		return
+	}
+	for _, dv := range dvList.Items {
+		if dv.UID == dataVolume.UID || getDeltaTTL(&dv, *dvTTL) > 0 {
+			continue
+		}
+		pvc := &corev1.PersistentVolumeClaim{}
+		if err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: dv.Namespace, Name: dv.Name}, pvc); err != nil {
+			return
+		}
+		if err = r.detachPvcDeleteDv(pvc, &dv); err != nil {
+			return
+		}
+	}
+	return &reconcile.Result{}, nil
+}
+
+func (r *DatavolumeReconciler) detachPvcDeleteDv(pvc *corev1.PersistentVolumeClaim, dv *cdiv1.DataVolume) error {
+	dvDelete := dv.Annotations[AnnDeleteAfterCompletion]
+	if IsSucceeded(pvc) && dvDelete == "true" {
+		updatePvcOwnerRefs(pvc, dv)
+		delete(pvc.Annotations, AnnPopulatedFor)
```

Comment:
Can there be more annotations we want to get rid of when detaching? `AnnPrePopulated`?

Reply:
`AnnPrePopulated` is on DV, `AnnPopulatedFor` is on PVC